services/wallet/concurrent.go

```diff
@@ -40,42 +45,202 @@ func (r *Result) Get() []Transfer {
 	GetTransfersByNumber(context.Context, *big.Int) ([]Transfer, error)
 }
 
-func downloadEthConcurrently(c *ConcurrentDownloader, client BalanceReader, downloader TransferDownloader, account common.Address, low, high *big.Int) {
+func checkRanges(parent context.Context, client BalanceReader, cache *balanceCache, downloader TransferDownloader, account common.Address, ranges [][]*big.Int) ([][]*big.Int, []*big.Int, error) {
+	ctx, cancel := context.WithTimeout(parent, 30*time.Second)
+	defer cancel()
+
+	c := NewConcurrentDownloader(ctx)
+
+	for _, blocksRange := range ranges {
+		from := blocksRange[0]
+		to := blocksRange[1]
+
+		c.Add(func(ctx context.Context) error {
+			if from.Cmp(to) >= 0 {
+				return nil
+			}
+			log.Debug("eth transfers comparing blocks", "from", from, "to", to)
+			lb, err := cache.BalanceAt(client, ctx, account, from)
+			if err != nil {
+				return err
+			}
+			hb, err := cache.BalanceAt(client, ctx, account, to)
+			if err != nil {
+				return err
+			}
+			if lb.Cmp(hb) == 0 {
+				log.Debug("balances are equal", "from", from, "to", to)
+				// In case if balances are equal but non zero we want to check if
+				// eth_getTransactionCount return different values, because there
+				// still might be transactions
+				if lb.Cmp(zero) != 0 {
+					return nil
+				}
+
+				ln, err := client.NonceAt(ctx, account, from)
+				if err != nil {
+					return err
+				}
+				hn, err := client.NonceAt(ctx, account, to)
+				if err != nil {
+					return err
+				}
+				if ln == hn {
+					log.Debug("transaction count is also equal", "from", from, "to", to)
+					return nil
+				}
+			}
+			if new(big.Int).Sub(to, from).Cmp(one) == 0 {
+				c.PushBlock(to)
+				return nil
+			}
+			mid := new(big.Int).Add(from, to)
+			mid = mid.Div(mid, two)
+			cache.BalanceAt(client, ctx, account, mid)
+			log.Debug("balances are not equal", "from", from, "mid", mid, "to", to)
+
+			c.PushRange([]*big.Int{from, mid})
+			c.PushRange([]*big.Int{mid, to})
+			return nil
+		})
+
+	}
+
+	select {
+	case <-c.WaitAsync():
+	case <-ctx.Done():
+		log.Error("eth downloader is stuck")
+		return nil, nil, errors.New("eth downloader is stuck")
+	}
+
+	if c.Error() != nil {
+		log.Error("failed to dowload transfers using concurrent downloader", "error", c.Error())
+		return nil, nil, c.Error()
+	}
+
+	return c.GetRanges(), c.GetBlocks(), nil
+}
+
+func findBlocksWithEthTransfers(parent context.Context, client BalanceReader, cache *balanceCache, downloader TransferDownloader, account common.Address, low, high *big.Int) (from *big.Int, blocks []*big.Int, err error) {
+	ranges := [][]*big.Int{{low, high}}
+	minBlock := big.NewInt(low.Int64())
+	blocks = []*big.Int{}
+	var lvl = 1
+	for len(ranges) > 0 && lvl <= 30 {
+		log.Info("check blocks ranges", "lvl", lvl, "ranges len", len(ranges))
+		lvl = lvl + 1
+		newRanges, newBlocks, err := checkRanges(parent, client, cache, downloader, account, ranges)
+		blocks = append(blocks, newBlocks...)
+		if err != nil {
```

Comment:
Shouldn't `err` be checked before appending?

Reply:
it should be, although `blocks` is "lost" in case of error anyway